pybossa/sched.py

```diff
@@ -292,24 +282,35 @@ def template_get_locked_task(project_id, user_id=None, user_ip=None,
     return template_get_locked_task
 
 
-@locked_scheduler
-def get_locked_task(
-    project_id,
-    user_id=None,
-    user_ip=None,
-    external_uid=None,
-    offset=0,
-    limit=1,
-    orderby='priority_0',
-    desc=True,
-    rand_within_priority=False,
-    present_gold_task=False,
-    gold_only=False
-):
-    having_clause = 'HAVING COUNT(task_run.task_id) < n_answers' if not (present_gold_task or gold_only) else ''
```

Comment:
@markroseberry This having clause was added here:

https://github.com/bloomberg/pybossa/commit/f976a6660eaff0e3d29e890c9d0dacdda9748599#diff-818837ad2f20f7b8415825710dde7de8R322

Was this added to address a specific issue?  Does the `task.state != 'completed'` do the same thing?  Also, I don't see a similar 'having clause' in the user preference scheduler.

All tests pass locally if it's removed from the query.

Reply:
Update: the tests are fixed now and that having clause was not needed, so please let me know.

Reply:
I didn't add the HAVING clause, I just modified it with the `gold_only` parameter. It was added in an earlier commit. The code in this module scares me. I can't say with confidence what is needed or not needed.

So, I would need to understand the specific reason the HAVING clause was added in the first place and then see if that reason still holds. Maybe it's just not being testing in the unit tests.

Reply:
I agree it's not clear.  However, all tests passed without it and I'd like to believe that if it was added to fix something, there would be a test for that problem (although may it wasn't easily reproducible).

However, that same clause was not part of the user preference scheduler...and this clause doesn't seem special to user preferences.  So if there were an underlying problem, it seems like it should have been added to both locked schedulers...

Maybe @dchhabda @fiorda can comment...they have much more experience here.

Reply:
Looks like @dchhabda added it: https://github.com/bloomberg/pybossa/commit/1229daf2886b12f8829017cbf34e3b3abc13cd9c